File: examples/py/websocket-subscribe-all.py
import os
import sys
import pprint
import traceback

pp = pprint.PrettyPrinter(depth=6)

root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(root + '/python')
# import ccxt  # noqa: E402
import ccxt.async_support as ccxt  # noqa: E402
import asyncio  # noqa: E402

loop = asyncio.get_event_loop()
# import txaio
# txaio.start_logging(level='debug')

def chunkIt(seq, num):
    avg = len(seq) / float(num)
    out = []
    last = 0.0

    while last < len(seq):
        out.append(seq[int(last):int(last + avg)])
        last += avg

    return out

async def main():
    if len(sys.argv) <= 2:
        print('python ' + __file__ + ' exchange apikey secret limit symbol ...')
        sys.exit(-1)

    exchange_id = sys.argv[1]
    apiKey = sys.argv[2]
    secret = sys.argv[3]
    limit = int(sys.argv[4])
    symbols = []
    eventSymbols = []
    for i in range(5, len(sys.argv)):
        symbols.append(sys.argv[i])
        eventSymbols.append({
            "event": "ob",
            "symbol": sys.argv[i],
            "params": {
                'limit': limit  
            }
        })

    exchange = getattr(ccxt, exchange_id)({
        "apiKey": apiKey,
        "secret": secret,
        "enableRateLimit": True,
        'verbose': False,
        'timeout': 5 * 1000,
        # 'wsproxy': 'http://185.93.3.123:8080/',
    })


    @exchange.on('err')
    def websocket_error(err, conxid):  # pylint: disable=W0612
        print(type(err).__name__ + ":" + str(err))
        traceback.print_tb(err.__traceback__)
        traceback.print_stack()
        loop.stop()

    @exchange.on('ob')
    def websocket_ob(symbol, ob):  # pylint: disable=W0612
        print("ob received from: " + symbol)
        sys.stdout.flush()
        # pp.pprint(ob)

    sys.stdout.flush()

    print("subscribe: " + ','.join(symbols))
    sys.stdout.flush()
    await exchange.websocket_subscribe_all(eventSymbols)
    print("subscribed: " + ','.join(symbols))
    sys.stdout.flush()
    await asyncio.sleep(10)

    chunkedSymbols = chunkIt(symbols,2)
    chunkedEventSymbols = chunkIt(eventSymbols,2)
    print("unsubscribe: " + ','.join(chunkedSymbols[1]))
    sys.stdout.flush()
    await exchange.websocket_unsubscribe_all(chunkedEventSymbols[1])
    print("unsubscribed: " + ','.join(chunkedSymbols[1]))
    await asyncio.sleep(10)
    print("unsubscribe: " + ','.join(chunkedSymbols[0]))
    sys.stdout.flush()
    await exchange.websocket_unsubscribe_all(chunkedEventSymbols[0])
    print("unsubscribed: " + ','.join(chunkedSymbols[0]))
    await asyncio.sleep(2)
    await exchange.close()



loop.run_until_complete(main())
# loop.run_forever()
# loop.stop()
# loop.close()
print("after complete")

File: examples/js/websocket-subscribe-all.js
'use strict';

const asTable = require('as-table'),
  log = require('ololog').noLocate,
  ansi = require('ansicolor').nice,
  ccxt = require('../../ccxt.js'),
  cex = require('../../js/cex');
//const HttpsProxyAgent = require('https-proxy-agent');

function printUsage() {
  log(
    'Usage: node',
    process.argv[1],
    'exchange',
    'apiKey',
    'secret',
    'limit',
    'symbol',
    '...'
  );
}
let sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

let exchange;
async function subscribeOrderBook(id, apiKey, secret, limit, symbols, eventSymbols) {
  // const proxy = process.env.http_proxy || 'http://localhost:7080'; // HTTP/HTTPS proxy to connect to（only local http proxy work at "here"）
  // const agent = new HttpsProxyAgent(proxy);
  exchange = new ccxt[id]({
    apiKey: apiKey,
    secret: secret,
    enableRateLimit: true,
    verbose: true,
    // agent: agent
  });
  exchange.on('err', (err, conxid) => {
    try {
      console.log(err);
      exchange.websocketClose(conxid);
    } catch (ex) {
      console.log(ex);
    }
  });
  exchange.on('ob', (market, ob) => {
    console.log('ob received: ', market, ob);
    // console.log (ob);
  });
  await exchange.loadMarkets();

  console.log('subscribe: ' + symbols.join(','));
  await exchange.websocketSubscribeAll(eventSymbols);
  console.log('subscribed: ' + symbols.join(','));
  await sleep(10 * 1000);

  console.log('unsubscribe: ' + symbols.slice(0,1).join(','));
  await exchange.websocketUnsubscribeAll(eventSymbols.slice(0,1));
  console.log('unsubscribed: ' + symbols.slice(0,1).join(','));
  await sleep(10 * 1000);

  console.log('unsubscribe: ' + symbols.slice(1).join(','));
  await exchange.websocketUnsubscribeAll(eventSymbols.slice(1));
  console.log('unsubscribed: ' + symbols.slice(1).join(','));
  await sleep(2 * 1000);

}

(async function main() {
  try {
    if (process.argv.length > 6) {
      const id = process.argv[2]
      const apiKey = process.argv[3]
      const secret = process.argv[4]
      const limit = parseInt (process.argv[5])
      const symbols = [];
      const eventSymbols = [];
      for (let i = 6 ; i < process.argv.length; i++) {
          symbols.push (process.argv[i].toUpperCase ())
          eventSymbols.push({
              "event": "ob",
              "symbol": process.argv[i],
              "params": {
                  'limit': limit  
              }
          });
      }
      //
      // this call gets live orderbook 
      const ob = await subscribeOrderBook(id, apiKey, secret, limit, symbols, eventSymbols);
    } else {
      printUsage ()
    }
  } catch (ex) {
    log('Error:'.red, ex);
    log(ex.stack);
    exchange.websocketClose();
  }
  // process.exit ()
})();
